CMakeLists.txt

```diff
@@ -1,81 +1,472 @@
+  set(CMAKE_C_FLAGS_DEBUG   "-O0 -g -ftrapuv -traceback")
+  set(CMAKE_C_FLAGS_RELEASE "-O2 -debug minimal -xCORE-AVX2")
+
+elseif(CMAKE_C_COMPILER_ID STREQUAL "IntelLLVM")
+
+  # Copied from MOM5/bin/mkmf.template.nci
+  set(CMAKE_C_FLAGS         "${CMAKE_C_FLAGS}")
+  set(CMAKE_C_FLAGS_DEBUG   "-O0 -g -ftrapuv -traceback")
+  set(CMAKE_C_FLAGS_RELEASE "-O2 -debug minimal -xCORE-AVX2")
 
-message("Using Fortran: ${CMAKE_Fortran_COMPILER_ID} ${CMAKE_Fortran_COMPILER} and C: ${CMAKE_C_COMPILER_ID} ${CMAKE_C_COMPILER}")
+elseif(CMAKE_C_COMPILER_ID MATCHES "Clang")
 
-if(CMAKE_Fortran_COMPILER_ID STREQUAL "Intel")
-  set(CMAKE_Fortran_FLAGS                "${CMAKE_Fortran_FLAGS} -fno-alias -stack-temps -safe-cray-ptr -ftz -assume byterecl -i4 -r8 -nowarn -sox -traceback")
-  set(CMAKE_Fortran_FLAGS_RELEASE        "${CMAKE_Fortran_FLAGS_RELEASE} -O2 -fp-model source")
-  set(CMAKE_Fortran_FLAGS_RELWITHDEBINFO "${CMAKE_Fortran_FLAGS_RELWITHDEBINFO} -g")
-  set(CMAKE_Fortran_FLAGS_DEBUG          "${CMAKE_Fortran_FLAGS_DEBUG} -g -O0 -check -check noarg_temp_created -check nopointer -warn -warn noerrors -fpe0 -ftrapuv")
-elseif(CMAKE_Fortran_COMPILER_ID STREQUAL "GNU")
-  set(CMAKE_Fortran_FLAGS                "${CMAKE_Fortran_FLAGS} -no-pie -fcray-pointer -fdefault-real-8 -ffree-line-length-none -fno-range-check")
-  set(CMAKE_Fortran_FLAGS_RELEASE        "${CMAKE_Fortran_FLAGS_RELEASE} -O2")
-  set(CMAKE_Fortran_FLAGS_RELWITHDEBINFO "${CMAKE_Fortran_FLAGS_RELWITHDEBINFO} -g")
-  set(CMAKE_Fortran_FLAGS_DEBUG          "${CMAKE_Fortran_FLAGS_DEBUG} -O0 -g -Wuninitialized -fcheck=bounds -Werror -ffpe-trap=invalid,zero,overflow")
-else ()
-  message ("Unknown FORTRAN compiler default flags only...")
+  # Copied from MOM5/bin/mkmf.template.nci.gfortran
+  set(CMAKE_C_FLAGS         "${CMAKE_C_FLAGS}")
+  set(CMAKE_C_FLAGS_DEBUG   "-O0 -g")
+  set(CMAKE_C_FLAGS_RELEASE "-O2")
+
+else()
+  message(WARNING "C compiler with ID ${CMAKE_C_COMPILER_ID} will be used with CMake default options")
 endif()
 
-if(CMAKE_C_COMPILER_ID STREQUAL "Intel")
-  set(CMAKE_C_FLAGS         "${CMAKE_C_FLAGS} -sox -traceback") 
-  set(CMAKE_C_FLAGS_DEBUG   "${CMAKE_C_FLAGS_DEBUG} -O0 -debug")
-  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2 -debug minimal")
-elseif(CMAKE_C_COMPILER_ID STREQUAL "GNU")
-  set(CMAKE_C_FLAGS         "${CMAKE_C_FLAGS} -frecord-gcc-switches")
-  set(CMAKE_C_FLAGS_DEBUG   "${CMAKE_C_FLAGS_DEBUG} -O0 -Werror -Wuninitialized -Wno-stringop-overflow")
-  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2")
-else ()
-  message ("Unknown C compiler default flags only...")
+################################################################################
+
+# Build options
+option(OPENMP      "Build FMS with OpenMP support"        OFF)
+option(32BIT       "Build 32-bit (r4) FMS library"        OFF)
+# Refer to CMAKE_Fortran_FLAGS above, it contains -r8
+option(64BIT       "Build 64-bit (r8) FMS library"        ON)
```

Comment:
Referring to the inline comment, should the setting of `-r8` be deferred and this option used to add that flag or not as is done with `r4_defs` below ? What does upstream do with the `-r8` flag?

As it is I suspect the `32BIT` version won't build with the `-r8` flag.

Reply:
Good question. Doesn't the compiler use the last compiler flag if there were preceding conflicting flags? I'll comment out the `32BIT` option, that is off by default, for now and we can enable it again if/when the compiler flags get updated.